File: src/train.py
# src/train.py
# Part 3 requirements:
# - Load train.csv (prefer Kaggle format)
# - Explore/adjust data with clear print logs (STEP 14–18)
# - Build a logistic regression model
# - Report training accuracy
# - Predict on test.csv and export submission (Kaggle test has no labels, so we print holdout accuracy from train split)
# - Work both locally and in Docker
# - Robust to weird CSV encodings/separators and column name variants

import pathlib
import sys
import pandas as pd
from joblib import dump
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder
from sklearn.impute import SimpleImputer
from sklearn.pipeline import Pipeline
from sklearn.metrics import accuracy_score
import csv

ROOT = pathlib.Path(__file__).resolve().parents[1]
DATA_DIR = ROOT / "src" / "data"
ART = ROOT / "artifacts"
ART.mkdir(exist_ok=True)

TRAIN = DATA_DIR / "train.csv"
TEST = DATA_DIR / "test.csv"
SINGLE = DATA_DIR / "titanic.csv"

EXPECTED_LOWER = ["survived", "pclass", "sex", "age", "sibsp", "parch", "fare", "embarked"]

def step(n, msg):
    print(f"[STEP {n}] {msg}", flush=True)

def ensure_data():
    have_kaggle = TRAIN.exists() or TEST.exists()
    have_single = SINGLE.exists()
    if not have_kaggle and not have_single:
        step(14, "No local data detected. Will run download_data.py to fetch a single-file fallback.")
        import subprocess, sys as _sys
        subprocess.check_call([_sys.executable, str(ROOT / "src" / "download_data.py")])
    else:
        present = [p.name for p in [TRAIN, TEST, SINGLE] if p.exists()]
        step(14, f"Detected local data files: {present}")

# ---------- Robust CSV reading + column normalization ----------

def _read_csv_robust(path):
    """
    Super-robust CSV reader:
    1) Try sep=None (sniff) with multiple encodings.
    2) If that fails, try explicit delimiters across encodings.
    3) If still failing, use csv.Sniffer() to guess delimiter per encoding.
    Supports weird Excel exports (UTF-16 + tab/comma) and pipe-delimited files.
    """
    import io, csv

    encodings = ["utf-8", "utf-8-sig", "cp1252", "latin1", "utf-16", "utf-16le", "utf-16be", "gbk"]
    seps = [None, ",", ";", "\t", "|"]   # None lets pandas sniff with python engine
    last_err = None

    # Try pandas sniffing first with multiple encodings
    for enc in encodings:
        try:
            df = pd.read_csv(path, encoding=enc, sep=None, engine="python", skipinitialspace=True)
            if df.shape[1] >= 3:
                return df
        except Exception as e:
            last_err = e

    # Try explicit delimiters x encodings
    for enc in encodings:
        for sep in seps[1:]:
            try:
                df = pd.read_csv(path, encoding=enc, sep=sep, engine="python", skipinitialspace=True)
                if df.shape[1] >= 3:
                    return df
            except Exception as e:
                last_err = e

    # Use csv.Sniffer on a small sample per encoding to pick delimiter, then read with pandas
    for enc in encodings:
        try:
            with open(path, "r", encoding=enc, errors="replace") as f:
                sample = f.read(4096)
            try:
                dialect = csv.Sniffer().sniff(sample, delimiters=";,|\t,")
                sep = dialect.delimiter
            except Exception:
                sep = ","  # fallback
            df = pd.read_csv(path, encoding=enc, sep=sep, engine="python", skipinitialspace=True)
            if df.shape[1] >= 3:
                return df
        except Exception as e:
            last_err = e
            continue

    raise RuntimeError(f"Failed to read {path} with extended encodings/separators. Last error: {last_err}")

def _normalize_columns(df: pd.DataFrame) -> pd.DataFrame:
    """
    Strip spaces, drop BOM, unify to lowercase, replace non-alnum with underscores,
    and fix common alias variants to expected names.
    """
    df.columns = (
        df.columns
        .str.strip()
        .str.replace("\ufeff", "", regex=False)          # BOM
        .str.replace(r"[^\w]+", "_", regex=True)         # non-alnum -> underscore
        .str.lower()
    )
    alias = {
        "p_class": "pclass",
        "ticket_class": "pclass",
        "sex_": "sex",
        "no_of_siblings_spouses": "sibsp",
        "siblings_spouses": "sibsp",
        "no_of_parents_children": "parch",
        "parents_children": "parch",
        "embark": "embarked",
    }
    df.rename(columns={k: v for k, v in alias.items() if k in df.columns}, inplace=True)
    return df

# ---------- Data loaders (Kaggle-preferred, fallback to single-file) ----------

def load_train_dataframe():
    if TRAIN.exists():
        step(15, "Using Kaggle train.csv as the training set (robust loader).")
        df = _read_csv_robust(TRAIN)
    elif SINGLE.exists():
        step(15, "Using single-file fallback titanic.csv as the training set (robust loader).")
        df = _read_csv_robust(SINGLE)
    else:
        print("[ERROR] No available training data.", file=sys.stderr)
        sys.exit(1)

    df = _normalize_columns(df)

    if "survived" not in df.columns:
        step(15, f"Columns detected: {list(df.columns)}")
        raise KeyError("Expected column 'Survived' not found after normalization.")

    keep_lower = ["survived", "pclass", "sex", "age", "sibsp", "parch", "fare", "embarked"]
    missing = [c for c in keep_lower if c not in df.columns]
    if missing:
        step(15, f"Columns detected: {list(df.columns)}")
        raise KeyError(f"Missing expected columns after normalization: {missing}")

    df = df[keep_lower].copy()
    # Rename back to TitleCase expected by downstream code
    df.rename(columns={
        "survived": "Survived",
        "pclass": "Pclass",
        "sex": "Sex",
        "age": "Age",
        "sibsp": "SibSp",
        "parch": "Parch",
        "fare": "Fare",
        "embarked": "Embarked"
    }, inplace=True)
    return df

def load_test_dataframe_like_train():
    if TEST.exists():
        step(16, "Loading Kaggle test.csv for predictions (robust loader).")
        df = _read_csv_robust(TEST)
        df = _normalize_columns(df)
        keep_lower = ["pclass", "sex", "age", "sibsp", "parch", "fare", "embarked"]
        missing = [c for c in keep_lower if c not in df.columns]
        if missing:
            step(16, f"Columns detected in test.csv: {list(df.columns)}")
            raise KeyError(f"Missing expected test columns after normalization: {missing}")
        df = df[keep_lower].copy()
        df.rename(columns={
            "pclass": "Pclass",
            "sex": "Sex",
            "age": "Age",
            "sibsp": "SibSp",
            "parch": "Parch",
            "fare": "Fare",
            "embarked": "Embarked"
        }, inplace=True)
        return df
    elif SINGLE.exists():
        step(16, "Single-file mode has no separate test.csv; will rely on holdout validation.")
        return None
    else:
        return None

# ---------- Model pipeline ----------

def build_pipeline():
    """
    Logistic regression with basic preprocessing:
    - Numeric: median imputation
    - Categorical: most_frequent imputation + OneHot
    """
    numeric = ["Age", "SibSp", "Parch", "Fare", "Pclass"]
    categorical = ["Sex", "Embarked"]

    pre = ColumnTransformer(
        transformers=[
            ("num", SimpleImputer(strategy="median"), numeric),
            ("cat", Pipeline([
                ("imp", SimpleImputer(strategy="most_frequent")),
                ("ohe", OneHotEncoder(handle_unknown="ignore"))
            ]), categorical),
        ],
        remainder="drop"
    )

    clf = LogisticRegression(max_iter=1000, n_jobs=None)
    pipe = Pipeline([
        ("pre", pre),
        ("clf", clf)
    ])
    return pipe, numeric + categorical

# ---------- Main ----------

def main():
    # === STEP 14: Data presence and environment check ===
    ensure_data()

    # === STEP 15: Light EDA / cleaning ===
    df = load_train_dataframe()
    step(15, f"Train shape: {df.shape}")
    step(15, f"Missing values (head):\n{df.isna().sum().head()}")

    y = df["Survived"].astype(int)
    X = df.drop(columns=["Survived"])

    # === STEP 16: Feature engineering and model assembly ===
    pipe, feature_cols = build_pipeline()
    step(16, f"Feature columns used: {feature_cols}")

    # === STEP 17: Fit on full train and report training accuracy ===
    step(17, "Fitting logistic regression (max_iter=1000).")
    pipe.fit(X, y)
    train_pred = pipe.predict(X)
    train_acc = accuracy_score(y, train_pred)
    step(17, f"Training accuracy: {train_acc:.4f}")

    # === STEP 18: Holdout validation from train (since Kaggle test has no labels) ===
    step(18, "Performing 80/20 holdout split on train.csv.")
    Xtr, Xte, ytr, yte = train_test_split(X, y, test_size=0.2, random_state=42, stratify=y)
    pipe2, _ = build_pipeline()
    pipe2.fit(Xtr, ytr)
    holdout_acc = accuracy_score(yte, pipe2.predict(Xte))
    step(18, f"Holdout accuracy: {holdout_acc:.4f}")

    # Save model
    model_path = ART / "model.joblib"
    dump(pipe2, model_path)
    step(18, f"Saved model to: {model_path}")

    # Predict on Kaggle test.csv if present and export submission
    test_df = load_test_dataframe_like_train()
    if test_df is not None and TEST.exists():
        step(18, "Generating predictions for Kaggle test.csv and writing artifacts/submission.csv.")
        sub_pred = pipe2.predict(test_df).astype(int)
        raw_test = pd.read_csv(TEST)
        submission = pd.DataFrame({"PassengerId": raw_test["PassengerId"], "Survived": sub_pred})
        out = ART / "submission.csv"
        submission.to_csv(out, index=False)
        step(18, f"Wrote submission file: {out} (Note: official test has no labels, so no test accuracy here.)")
    else:
        step(18, "Kaggle test.csv not found. Skipping submission export.")

    print("\n[SUMMARY]")
    print(f"  Train accuracy : {train_acc:.4f}")
    print(f"  Holdout accuracy: {holdout_acc:.4f}")
    if (ART / "submission.csv").exists():
        print(f"  Submission path : {ART / 'submission.csv'}")
    print(f"  Model path      : {model_path}")

if __name__ == "__main__":
    main()

File: src/r/train.R
# src/r/train.R
# Part 4 (R) — Repeat steps 14–21 with clear console logs.
s <- function(n, msg) cat(sprintf("[STEP %d] %s\n", n, msg))

s(14, "R runtime starting...")

suppressPackageStartupMessages({
  library(readr)
  library(dplyr)
  library(stringr)
  library(tibble)
})

root <- normalizePath("/app", mustWork = FALSE)
data_dir <- file.path(root, "src", "data")
art_dir  <- file.path(root, "artifacts")

if (!dir.exists(art_dir)) dir.create(art_dir, recursive = TRUE, showWarnings = FALSE)

train_path <- file.path(data_dir, "train.csv")
test_path  <- file.path(data_dir, "test.csv")

if (!file.exists(train_path)) {
  s(14, sprintf("train.csv not found at %s", train_path))
  stop("No train.csv; please place Kaggle files under src/data/")
} else {
  s(14, sprintf("Detected data files: %s%s",
                basename(train_path),
                if (file.exists(test_path)) paste0(", ", basename(test_path)) else ""))
}

# ---- STEP 15: Load and light EDA / cleaning ----
read_csv_robust <- function(path) {
  # Try comma first (Kaggle default), then semicolon, then tab
  tries <- list(
    function() readr::read_csv(path, show_col_types = FALSE, progress = FALSE),
    function() readr::read_csv2(path, show_col_types = FALSE, progress = FALSE),
    function() readr::read_delim(path, delim = "\t", show_col_types = FALSE, progress = FALSE)
  )
  last_err <- NULL
  for (f in tries) {
    out <- tryCatch(f(), error = function(e) e)
    if (!inherits(out, "error")) return(out)
    last_err <- out
  }
  stop(sprintf("Failed to read %s: %s", path, as.character(last_err)))
}

s(15, "Loading Kaggle train.csv")
df <- read_csv_robust(train_path)

# Normalize column names (strip, lowercase)
norm_names <- function(x) {
  x |>
    str_replace_all("\\ufeff", "") |>
    str_trim() |>
    str_replace_all("[^A-Za-z0-9]+", "_") |>
    tolower()
}
names(df) <- norm_names(names(df))

# Minimal required columns
need <- c("survived","pclass","sex","age","sibsp","parch","fare","embarked")
missing <- setdiff(need, names(df))
if (length(missing)) {
  s(15, paste("Detected columns:", paste(names(df), collapse=", ")))
  stop(paste("Missing columns:", paste(missing, collapse=", ")))
}
df <- df[, need]

s(15, sprintf("Train shape: (%d, %d)", nrow(df), ncol(df)))
s(15, "Missing values (head):")
print(head(colSums(is.na(df))))

# ---- STEP 16: Feature engineering ----
# Impute simple medians / modes; ensure proper types
num_cols <- c("age","sibsp","parch","fare","pclass")
cat_cols <- c("sex","embarked")

# numeric impute
for (c in num_cols) {
  if (anyNA(df[[c]])) {
    med <- median(df[[c]], na.rm = TRUE)
    df[[c]][is.na(df[[c]])] <- med
    s(16, sprintf("Imputed median for %s = %.4f", c, med))
  }
}
# categorical impute
for (c in cat_cols) {
  if (anyNA(df[[c]])) {
    mode <- names(sort(table(df[[c]]), decreasing = TRUE))[1]
    df[[c]][is.na(df[[c]])] <- mode
    s(16, sprintf("Imputed mode for %s = %s", c, mode))
  }
}

# factors for GLM; Survived numeric 0/1
df <- df |>
  mutate(
    survived = as.integer(survived),
    sex = factor(sex),
    embarked = factor(embarked),
    pclass = as.integer(pclass)
  )

s(16, "Feature columns used: pclass, sex, age, sibsp, parch, fare, embarked")

# ---- STEP 17: Fit on full train and report training accuracy ----
s(17, "Fitting logistic regression (glm, binomial).")
glm_full <- glm(
  survived ~ pclass + sex + age + sibsp + parch + fare + embarked,
  data = df, family = binomial()
)
pred_train_prob <- as.numeric(predict(glm_full, type = "response"))
pred_train <- as.integer(pred_train_prob >= 0.5)
train_acc <- mean(pred_train == df$survived)
s(17, sprintf("Training accuracy: %.4f", train_acc))

# ---- STEP 18: Holdout validation (80/20 split on train) ----
set.seed(42)
idx <- sample(seq_len(nrow(df)), size = floor(0.8 * nrow(df)))
tr <- df[idx, , drop = FALSE]
te <- df[-idx, , drop = FALSE]
glm_hold <- glm(
  survived ~ pclass + sex + age + sibsp + parch + fare + embarked,
  data = tr, family = binomial()
)
pred_hold <- as.numeric(predict(glm_hold, newdata = te, type = "response"))
hold_acc <- mean(as.integer(pred_hold >= 0.5) == te$survived)
s(18, sprintf("Holdout accuracy: %.4f", hold_acc))

# ---- STEP 19: Persist model artifacts ----
model_rds <- file.path(art_dir, "model_r_glm.rds")
saveRDS(glm_hold, model_rds)
s(19, sprintf("Saved R model to: %s", model_rds))

# ---- STEP 20: Predict test.csv (if present) and export submission ----
if (file.exists(test_path)) {
  s(20, "Loading Kaggle test.csv and generating submission.")
  test <- read_csv_robust(test_path)
  names(test) <- norm_names(names(test))
  # keep needed features, impute using train med/mode
  need_te <- c("pclass","sex","age","sibsp","parch","fare","embarked","passengerid")
  miss_te <- setdiff(need_te, names(test))
  if (length(miss_te)) {
    s(20, paste("Detected test columns:", paste(names(test), collapse=", ")))
    stop(paste("Missing test columns:", paste(miss_te, collapse=", ")))
  }
  test <- test[, need_te]
  
  # impute using train statistics
  # numeric
  for (c in c("age","sibsp","parch","fare","pclass")) {
    med <- median(df[[c]], na.rm = TRUE)
    test[[c]][is.na(test[[c]])] <- med
  }
  # categorical
  for (c in c("sex","embarked")) {
    mode <- names(sort(table(df[[c]]), decreasing = TRUE))[1]
    test[[c]][is.na(test[[c]])] <- mode
  }
  
  test <- test |>
    mutate(
      sex = factor(sex, levels = levels(df$sex)),
      embarked = factor(embarked, levels = levels(df$embarked)),
      pclass = as.integer(pclass)
    )
  
  prob <- as.numeric(predict(glm_hold, newdata = test, type = "response"))
  pred <- as.integer(prob >= 0.5)
  submission <- tibble(PassengerId = test$passengerid, Survived = pred)
  sub_path <- file.path(art_dir, "submission_r.csv")
  readr::write_csv(submission, sub_path)
  s(20, sprintf("Wrote submission: %s (Note: official test has no labels)", sub_path))
} else {
  s(20, "No test.csv detected; skipping submission.")
}

# ---- STEP 21: Summary ----
cat("\n[SUMMARY - R]\n")
cat(sprintf("  Train accuracy  : %.4f\n", train_acc))
cat(sprintf("  Holdout accuracy: %.4f\n", hold_acc))
cat(sprintf("  Model path      : %s\n", model_rds))
if (file.exists(file.path(art_dir, "submission_r.csv"))) {
  cat(sprintf("  Submission path : %s\n", file.path(art_dir, "submission_r.csv")))
}
